### ISSUE TITLE
Shock Rifle Primary Fires Extra Shot When Spammed in Network Matches

### ISSUE THREAD

**Issue by Barktooth** [OldUnreal/UnrealTournamentPatches#644]
Greetings. This is my first time reporting a bug here so I hope I'm not doing anything wrong. I did search the issue tracker and wasn't able to find anything that seemed related. The issue can be replicated using the following steps:
1. Join a networked game on a server using stock UT netcode, or Ultimate NewNet (but not Frag NewNet).
2. Pick up a Shock Rifle or Enhanced Shock Rifle.
3. Mash the primary fire button as fast as possible for a few seconds.
4. Cease pressing the button right after the weapon has fired a shot.

Expected result: weapon stops firing.
Actual result: weapon fires one extra shot a full reload after you have stopped spamming the fire button.

Interestingly, the same does not happen with the Enhanced Shock Rifle's secondary mode. No matter how much you spam it, it behaves as expected. Shock Rifle secondary also works correctly, and so do other weapons from my brief testing, although, anecdotally, I feel as though I've experienced the same bug with Flak Cannon before, but I'm unable to replicate it in testing. The bug is not present in offline practice matches, and as I alluded to above, is also seemingly fixed in Frag NewNet (v.136). [Here ](https://www.youtube.com/watch?v=w-GqNWzOl_I)is a video of me demonstrating the bug on several servers; if you pay attention to the mouse input overlay in the bottom right corner, you will notice that the weapon fires long after I have stopped clicking. The first server is regular netcode with some mutators that might affect things, but the second one is seemingly a completely stock server and the bug happens there as well. The third server is running Ultimate NewNet and the bug still happens. The fourth server has Frag NewNet and I wasn't able to trigger the bug there. 

Here's my setup:
Windows 10 64-bit version 21H1 Build 19043.1415
UT99 v469b Release (though I'm fairly sure the bug was present when I used to use Windows 7 and v436 too)
No mods

Some other notes:
Binding Fire to a key on the keyboard results in the same behaviour.
Playing around with the FPS cap does not seem to affect anything.
Changing netspeed doesn't affect it either.
Ping being low or high does not seem to matter.

**Comment by uralets** [comment]
I might be mixing up things, but I have a feeling I've experienced this behavior long before 469 (and probably on Linux too).

**Comment by Engineer9736** [comment]
Yeah i think UT always did this already. I remember when you're fragging, and trying to shoot again as fast as possible, you always have to wait for the weapon to become recharged for like 1 second; At some point you decide to wait a little so you can time the shot exactly with the crosshair being on the target, at that moment the weapon always fires one time more even though you didn't touch the mouse button for like 2+ seconds.

It would be great improvement indeed if it didn't do that. It's often in that second of waiting for the gun to stop shooting that you're getting shot. Although, it will be a massive gameplay change, i'm not sure how the die-hards think about such change (don't know if the real UT die-hards still exist though lol).

For me it would be a good change 👍

**Comment by SeriousBuggie** [comment]
This behavior created intentionally:
```
state ClientFiring
{
	simulated function bool ClientFire(float Value)
	{
		if ( Level.TimeSeconds - TapTime < 0.2 )
			return false;
		bForceFire = bForceFire || ( bCanClientFire && (Pawn(Owner) != None) && (AmmoType.AmmoAmount > 0) );
		return bForceFire;
	}

	simulated function bool ClientAltFire(float Value)
	{
		if ( Level.TimeSeconds - TapTime < 0.2 )
			return false;
		bForceAltFire = bForceAltFire || ( bCanClientFire && (Pawn(Owner) != None) && (AmmoType.AmmoAmount > 0) );
		return bForceAltFire;
	}

	simulated function AnimEnd()
	{
		local bool bForce, bForceAlt;

		bForce = bForceFire;
		bForceAlt = bForceAltFire;
		bForceFire = false;
		bForceAltFire = false;

		if ( bCanClientFire && (PlayerPawn(Owner) != None) && (AmmoType.AmmoAmount > 0) )
		{
			if ( bForce || (Pawn(Owner).bFire != 0) )
			{
				Global.ClientFire(0);
				return;
			}
			else if ( bForceAlt || (Pawn(Owner).bAltFire != 0) )
			{
				Global.ClientAltFire(0);
				return;
			}
		}			
		Super.AnimEnd();
	}
```
It use delayed bForceFire and bForceAltFire vars for make this possible.

**Comment by SeriousBuggie** [comment]
State `ClientFiring` used only on net and only for client. In `TournamentWeapon`:
```
simulated function bool ClientFire( float Value )
{
	if ( bCanClientFire && ((Role == ROLE_Authority) || (AmmoType == None) || (AmmoType.AmmoAmount > 0)) )
	{
		if ( (PlayerPawn(Owner) != None) 
			&& ((Level.NetMode == NM_Standalone) || PlayerPawn(Owner).Player.IsA('ViewPort')) )
		{
			if ( InstFlash != 0.0 )
				PlayerPawn(Owner).ClientInstantFlash( InstFlash, InstFog);
			PlayerPawn(Owner).ShakeView(ShakeTime, ShakeMag, ShakeVert);
		}
		if ( Affector != None )
			Affector.FireEffect();
		PlayFiring();
		if ( Role < ROLE_Authority )
			GotoState('ClientFiring');
		return true;
	}
	return false;
}		
```
By default this state look like this:
```
state ClientFiring
{
	simulated function bool ClientFire(float Value)
	{
		return false;
	}

	simulated function bool ClientAltFire(float Value)
	{
		return false;
	}

	simulated function AnimEnd()
	{
		if ( (Pawn(Owner) == None)
			|| ((AmmoType != None) && (AmmoType.AmmoAmount <= 0)) )
		{
			PlayIdleAnim();
			GotoState('');
		}
		else if ( !bCanClientFire )
			GotoState('');
		else if ( Pawn(Owner).bFire != 0 )
			Global.ClientFire(0);
		else if ( Pawn(Owner).bAltFire != 0 )
			Global.ClientAltFire(0);
		else
		{
			PlayIdleAnim();
			GotoState('');
		}
	}

	simulated function EndState()
	{
		AmbientSound = None;
	}
}
```
For shock rifle it override with code above.,
Also exists state `ClientAltFiring` which triggered in same way:
```
simulated function bool ClientAltFire( float Value )
{
	if ( bCanClientFire && ((Role == ROLE_Authority) || (AmmoType == None) || (AmmoType.AmmoAmount > 0)) )
	{
		if ( (PlayerPawn(Owner) != None) 
			&& ((Level.NetMode == NM_Standalone) || PlayerPawn(Owner).Player.IsA('ViewPort')) )
		{
			if ( InstFlash != 0.0 )
				PlayerPawn(Owner).ClientInstantFlash( InstFlash, InstFog);
			PlayerPawn(Owner).ShakeView(ShakeTime, ShakeMag, ShakeVert);
		}
		if ( Affector != None )
			Affector.FireEffect();
		PlayAltFiring();
		if ( Role < ROLE_Authority )
			GotoState('ClientAltFiring');
		return true;
	}
	return false;
}
```
And it default code:
```
state ClientAltFiring
{
	simulated function bool ClientFire(float Value)
	{
		return false;
	}

	simulated function bool ClientAltFire(float Value)
	{
		return false;
	}
	simulated function AnimEnd()
	{
		if ( (Pawn(Owner) == None)
			|| ((AmmoType != None) && (AmmoType.AmmoAmount <= 0)) )
		{
			PlayIdleAnim();
			GotoState('');
		}
		else if ( !bCanClientFire )
			GotoState('');
		else if ( Pawn(Owner).bFire != 0 )
			Global.ClientFire(0);
		else if ( Pawn(Owner).bAltFire != 0 )
			Global.ClientAltFire(0);
		else
		{
			PlayIdleAnim();
			GotoState('');
		}
	}

	simulated function EndState()
	{
		AmbientSound = None;
	}
}
```
But not override in shock.
Default code for state `ClientFiring` and `ClientAltFiring` is 100% same.

And in v436 override state code for shock exactly same.

So, because `ClientAltFiring` not override, then issue can be only after animation end for primary fire.
You can make two clicks - first primary, and next alt. On net you see remembered shot. Same if second click be for primary shot.
If you make both - primary shot "wins" and use only it.

Not sure for what this change introduced,

Also there exists special cool down delay for 0.2 second, for ignore any shoot right after enter to this state.

**Comment by SeriousBuggie** [comment]
As another weird test:
Make 2 primary clicks and hold alt fire.
Appear two primary shots, only after that start spam alt fire.

As summarize:
**Shock rifle, only on network, intentionally use modified code, which remember what player press after 0.2 seconds after primary fire, and make one additional shot with priority for primary fire.**

IDK why or for what this code written. But usually any code appear as answer for some problem. Need know which problem EPIC try solve here for say something more or possible "fix" this code,

Just note: if you use hold primary fire - no such issue. Maybe EPIC not count for constant spam of primary fire? 
It is not typical usage for weapon.

Also this is not a copy/paste bug, because such code is unique. Most close code located in bio rifle, but it is very different:
```
// Finish a firing sequence
function Finish()
{
	local bool bForce, bForceAlt;

	bForce = bForceFire;
	bForceAlt = bForceAltFire;
	bForceFire = false;
	bForceAltFire = false;

	if ( bChangeWeapon )
		GotoState('DownWeapon');
	else if ( PlayerPawn(Owner) == None )
	{
		Pawn(Owner).bAltFire = 0;
		Super.Finish();
	}
	else if ( (AmmoType.AmmoAmount<=0) || (Pawn(Owner).Weapon != self) )
		GotoState('Idle');
	else if ( (Pawn(Owner).bFire!=0) || bForce )
		Global.Fire(0);
	else if ( (Pawn(Owner).bAltFire!=0) || bForceAlt )
		Global.AltFire(0);
	else 
		GotoState('Idle');
}
```

**Comment by SeriousBuggie** [comment]
in v400 used exactly same code as now.
in v348, almost same, except delay for 0.2 seconds:
```
state ClientFiring
{
	simulated function bool ClientFire(float Value)
	{
		bForceFire = bForceFire || ( bCanClientFire && (Pawn(Owner) != None) && (AmmoType.AmmoAmount > 0) );
		return bForceFire;
	}

	simulated function bool ClientAltFire(float Value)
	{
		bForceAltFire = bForceAltFire || ( bCanClientFire && (Pawn(Owner) != None) && (AmmoType.AmmoAmount > 0) );
		return bForceAltFire;
	}

	simulated function AnimEnd()
	{
		local bool bForce, bForceAlt;

		bForce = bForceFire;
		bForceAlt = bForceAltFire;
		bForceFire = false;
		bForceAltFire = false;

		if ( bCanClientFire && (PlayerPawn(Owner) != None) && (AmmoType.AmmoAmount > 0) )
		{
			if ( bForce || (Pawn(Owner).bFire != 0) )
			{
				Global.ClientFire(0);
				return;
			}
			else if ( bForceAlt || (Pawn(Owner).bAltFire != 0) )
			{
				Global.ClientAltFire(0);
				return;
			}
		}			
		Super.AnimEnd();
	}

	simulated function EndState()
	{
		bForceFire = false;
		bForceAltFire = false;
	}

	simulated function BeginState()
	{
		bForceFire = false;
		bForceAltFire = false;
	}
}
```
So now we at least know 2 things:
1. It is not mistake, because if it be mistake, on modify code it will be fixed.
2. 0.2 sec timeout appear later and solve or another problem or try improve solve original problem.

**Comment by Barktooth** [comment]
> For shock rifle it override with code above.,

Oh I see, so that's why it does that only for Shock. Weird that they wrote but didn't use the override for alt fire, but maybe it ended up not being needed to fix whatever issue they were trying to address with this code. I have little idea why this would be introduced, my guess would be something related to the Shock combo functionality since that's the "special" thing about Shock Rifle compared to other weapons, but that's a complete shot in the dark. In any case, should I close this topic since it seems unlikely that this is worth looking into any more? It's a fairly minor issue anyway since as you said spam-clicking is not typical usage, I just manage to trigger this extra shot in heated firefights sometimes.

**Comment by SeriousBuggie** [comment]
v222 not use this fix.
v322(3dfx) not use this fix.
v338 use this fix.
```
      Unreal Tournament Demo Changes in Version 338 since 3dfx demo
```
```
  - improved client side weapon animation (RL and shockrifle)  
```
not sure if related:
```
  - fixed bots not playing rapid fire animation
```
```
  - Added first person demo recording/playback in botmatches  
```
About delay for 0.2 sec:
```
Version 348: Changes since version 338
```
No any exact info.

**Comment by SeriousBuggie** [comment]
Let's see what @stijn-volckaert say.
This behavior look totally weird.  Net play must act same as standalone (or listen) wherever this possible.

But before do something need know reason for introduce such code for not make things worse.

**Comment by Barktooth** [comment]
Wow, that's some great investigative work there, SeriousBuggie! Looks like it might have something to do with animations, which I didn't even consider. Waiting for stijn-volckaert seems like a good idea; I suppose I'll leave the topic open, then.

**Comment by stijn-volckaert** [comment]
I agree that this behavior is weird and I cannot really see the point of it. Unfortunately, I do not think we can fix this without breaking a bunch of mods that rely on the current behavior. I've marked this as wontfix for now, but we might revisit this issue later since we're currently working on a "compatibility package" for the upcoming 469c patch. This package will contain various fixes that have potential side effects. Thus, rather than fixing the base class, we're giving mappers and modders the option to use the fixed actors in their maps/mods. This package will also work for earlier clients.

**Comment by SeriousBuggie** [comment]
Replace weapon, even with subclass, not safe operation.
It is lead for next problems:
1. Weapons priority - broken.
2. Bind scheme which most use `GetWeapon` - broken.
3. Mutators which try replace weapon - broken.
ShockRifle very special case, because exists subclass SuperShockRifle. So nobody can use `IsA('ShockRifle')`. All will use `Class.Name == 'ShockRifle'`.

So replace weapon not an option. We not touch it, or make special fix via flags, which can be enabled or disabled. No other way.
Good news, it is really easy done via flags.
Bad news, this code on client side, so this fix work only for new network clients.

Exists one workaround - external actor in special package which every tick set
```
bForceFire = false;
bForceAltFire = false;
```
for any shock rifle owned by client on client side.
Look dirty, but do job. Impact for performance must be not big.
Work for all client. Can be controlled from server for all clients.

Look like acceptable solution.

**Comment by Engineer9736** [comment]
I've made a quick&dirty mutator that sets these variables to false every tick. Though the weapon still shoots 1 time too much, only without the fireball being shown.

The weapon still moves and makes the shooting sound. So there probably needs to be hacked some more things to get the desired result.

I've tried both simulated function tick and non-simulated function tick; it behaves the same.

Feel free to fork & further mess around with what i made.

https://github.com/Engineer9736/StopShockRifleShootingTooMuch/blob/master/Classes/StopShockRifleShootingTooMuch.uc

**Comment by SeriousBuggie** [comment]
It must be client side actor. So must be replicated to client at least. So appear in ServerPackages or embed to map.
Also not sure if mutators replicated to client side by default.
So better check via embed actor in map.

**Comment by Engineer9736** [comment]
Yeah i loaded it as a ServerActor and also put it in ServerPackages, but with this result. I'll look later into it, i'm supposed to be busy for work currently instead of making UT mods 😅

**Comment by SeriousBuggie** [comment]
```
class FixShock expands Keypoint;

simulated function Tick(float delta) {
	local shockRifle SR;
	Foreach AllActors(class'ShockRifle', SR) {
//		if (SR.bForceFire) Log(SR @ "bForceFire");
		SR.bForceFire = false;
//		if (SR.bForceAltFire) Log(SR @ "bForceAltFire");
		SR.bForceAltFire = false;
	}
}

defaultproperties
{
	bStatic=False
	bNoDelete=True
	RemoteRole=ROLE_SimulatedProxy
}
```
Work flawless.

look like you forget set `RemoteRole=ROLE_SimulatedProxy` and on client `Tick` not called.
